### PR TITLE
Mongo store extension and fixed delay schedule

### PR DESCRIPTION
**Pull request by ankitnanglia**

### CODE REVIEW

File: extensions/jdbc-store/src/main/java/com/cognitree/kronos/scheduler/store/jdbc/StdJDBCStoreService.java [hunk 1]
```diff
@@ -92,7 +92,6 @@ public void init() throws Exception {
     public void start() throws Exception {
         logger.info("Starting JDBC store service");
         initDatabase();
-        ServiceProvider.registerService(this);
```
Comment (girishkarthik) [inline extensions/jdbc-store/src/main/java/com/cognitree/kronos/scheduler/store/jdbc/StdJDBCStoreService.java]: Why is this removed?

File: extensions/mongo-store/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+```
+
+Here, A [MongoStoreService](src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoStoreService.java) is configured.
+Other optional configurable parameters are
+    - host (ip address of the machine running mongodb, defaults to `localhost`)
+    - port (port of the mongodb, defaults to `27017`)
+    - user (username to use for authentication)
+    - password (password of the user to use for authentication)
+    - authDatabase (authentication database to use for authentication)
+    - quartzDatabase (name of the database to use for storing quartz state)
```
Comment (girishkarthik) [inline extensions/mongo-store/README.md]: Do we have to expose "quartz" here? We mentioned above as Kronos state.

File: extensions/mongo-store/pom.xml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+                </exclusion>
+                <exclusion>
+                    <groupId>org.quartz-scheduler</groupId>
+                    <artifactId>quartz</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>com.cognitree.kronos</groupId>
+            <artifactId>jdbc-store</artifactId>
```
Comment (girishkarthik) [inline extensions/mongo-store/pom.xml]: Why do we need JDBC store for mongo-db store?

File: extensions/mongo-store/pom.xml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.mongodb</groupId>
+            <artifactId>mongo-java-driver</artifactId>
+            <version>3.10.1</version>
+        </dependency>
+        <dependency>
+            <groupId>com.novemberain</groupId>
+            <artifactId>quartz-mongodb</artifactId>
+            <version>2.1.0</version>
```
Comment (girishkarthik) [inline extensions/mongo-store/pom.xml]: Can we move versions as constants/properties?

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+                    gt("createdAt", createdAfter))));
+            pipelines.add(Aggregates.group("$status", Accumulators.sum("count", 1)));
+
+            AggregateIterable<Document> jobs = mongoCollection.aggregate(pipelines);
+            HashMap<Job.Status, Integer> statusMap = new HashMap<>();
+            for (Document job : jobs) {
+                Job.Status status = Job.Status.valueOf(job.get("_id").toString());
+                statusMap.put(status, job.getInteger("count"));
+            }
+            logger.debug("loadByStatus : returned {} jobs for namespace {} created after {} created before {}",
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java]: the comment is "countByStatus" and not "loadByStatus".

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+            return statusMap;
+        } catch (Exception e) {
+            logger.error("Error counting jobs by status under namespace {}, created after {}, created before {}",
+                    namespace, createdAfter, createdBefore, e);
+            throw new StoreException(e.getMessage(), e.getCause());
+        }
+    }
+
+    @Override
+    public Map<Job.Status, Integer> countByStatusForWorkflowName(String namespace, String workflowName,
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java]: Can this method reuse logic from above method "countByStatus"

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+                    and(
+                            eq("namespace", namespace),
+                            eq("workflow", workflowName)));
+        } catch (Exception e) {
+            logger.error("Error deleting jobs with workflow name {}, namespace {}", workflowName, namespace, e);
+            throw new StoreException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private MongoCollection<Job> getMongoCollection(String namespace) {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java]: Can this method be renamed as "getJobCollection" than "getMongoCollection" ?

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        }
+    }
+
+    @Override
+    public Map<Job.Status, Integer> countByStatusForWorkflowName(String namespace, String workflowName,
+                                                                 long createdAfter, long createdBefore) throws StoreException {
+        logger.debug("Received request to count by status having workflow name {}, namespace {}, created after {}, created before {}",
+                workflowName, namespace, createdAfter, createdBefore);
+        try {
+            MongoCollection<Document> mongoCollection = mongoClient.getDatabase(namespace)
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java]: Is this a Document or can be cast as "Jobs" as in the rest of the cases?

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoNamespaceStore.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            MongoDatabase database = mongoClient.getDatabase(namespaceId.getName());
+            database.drop();
+            getMongoCollection().deleteOne(eq("name", namespaceId.getName()));
+        } catch (Exception e) {
+            logger.error("Error deleting namespace with id {}", namespaceId, e);
+            throw new StoreException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private MongoCollection<Namespace> getMongoCollection() {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoNamespaceStore.java]: Can this method be renamed as getNamespaceCollection()

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoStoreService.java [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+/**
+ * A standard MongoDB connection based implementation of {@link StoreService}.
+ */
+public class MongoStoreService extends StoreService {
+
+    private static final Logger logger = LoggerFactory.getLogger(MongoStoreService.class);
+
+    private static final String HOST = "host";
+    private static final String DEFAULT_HOST = "localhost";
+    private static final String PORT = "port";
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoStoreService.java]: Can we append "_KEY" to all constants that represent a key or a property?

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        try {
+            MongoCollection<Task> mongoCollection = getMongoCollection(taskId.getNamespace());
+            mongoCollection.deleteOne(eq("job", taskId.getJob()));
+        } catch (Exception e) {
+            logger.error("Error deleting task with id {}", taskId, e);
+            throw new StoreException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private MongoCollection<Task> getMongoCollection(String namespace) {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java]: Can this be renamed as getTaskCollection?

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+import static com.mongodb.client.model.Filters.gt;
+import static com.mongodb.client.model.Filters.in;
+import static com.mongodb.client.model.Filters.lt;
+import static com.mongodb.client.model.Updates.combine;
+import static com.mongodb.client.model.Updates.set;
+
+/**
+ * A standard MongoDB based implementation of {@link TaskStore}.
+ */
+public class MongoTaskStore implements TaskStore {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java]: Can this class be merged with MongoJobStore with Generics.. ? Looks like lot of boilerplate code.

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowStore.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                    and(
+                            eq("name", workflowId.getName()),
+                            eq("namespace", workflowId.getNamespace())));
+        } catch (Exception e) {
+            logger.error("Error deleting workflow with id {}", workflowId, e);
+            throw new StoreException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private MongoCollection<Workflow> getMongoCollection(String namespace) {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowStore.java]: Method rename.

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowStore.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+import static com.mongodb.client.model.Updates.combine;
+import static com.mongodb.client.model.Updates.set;
+
+/**
+ * A standard MongoDB based implementation of {@link WorkflowStore}.
+ */
+public class MongoWorkflowStore implements WorkflowStore {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowStore.java]: Bolierplate code. Consider refactoring with MongoTaskStore and MongoJobStore.

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowTriggerStore.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+                            eq("name", workflowTriggerId.getName()),
+                            eq("workflow", workflowTriggerId.getWorkflow()),
+                            eq("namespace", workflowTriggerId.getNamespace())));
+        } catch (Exception e) {
+            logger.error("Error delete workflow trigger with id {}", workflowTriggerId, e);
+            throw new StoreException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private MongoCollection<WorkflowTrigger> getMongoCollection(String namespace) {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowTriggerStore.java]: Method rename.

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowTriggerStore.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+import static com.mongodb.client.model.Updates.combine;
+import static com.mongodb.client.model.Updates.set;
+
+/**
+ * A standard MongoDB based implementation of {@link WorkflowTrigger}.
+ */
+public class MongoWorkflowTriggerStore implements WorkflowTriggerStore {
```
Comment (girishkarthik) [inline extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoWorkflowTriggerStore.java]: Bolierplate code. Consider refactoring with MongoTaskStore and MongoJobStore.

File: scheduler/src/main/java/com/cognitree/kronos/scheduler/SchedulerApp.java [hunk 17]
```diff
@@ -91,8 +91,8 @@ public void start() throws Exception {
         jobService.start();
         workflowTriggerService.start();
         mailService.start();
-        workflowSchedulerService.start();
         taskSchedulerService.start();
+        workflowSchedulerService.start();
```
Comment (girishkarthik) [inline scheduler/src/main/java/com/cognitree/kronos/scheduler/SchedulerApp.java]: Add a comment as to why workflowschedulerservice should start after taskschedulerservice to avoid re-ordering later.

File: scheduler/src/main/java/com/cognitree/kronos/scheduler/WorkflowSchedulerService.java [hunk 18]
```diff
@@ -113,7 +115,7 @@ private void addJob(Workflow workflow, boolean replace) throws SchedulerExceptio
         scheduler.addJob(jobDetail, replace);
     }
 
-    void add(WorkflowTrigger workflowTrigger)
+    synchronized void add(WorkflowTrigger workflowTrigger)
```
Comment (girishkarthik) [inline scheduler/src/main/java/com/cognitree/kronos/scheduler/WorkflowSchedulerService.java]: Can we add the reason for synchronized keyword?

File: scheduler/src/main/java/com/cognitree/kronos/scheduler/model/FixedDelaySchedule.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
+package com.cognitree.kronos.scheduler.model;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+
+import java.util.Objects;
+
+import static com.cognitree.kronos.scheduler.model.Schedule.Type.fixed;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class FixedDelaySchedule extends Schedule {
```
Comment (girishkarthik) [inline scheduler/src/main/java/com/cognitree/kronos/scheduler/model/FixedDelaySchedule.java]: Class level comment maybe a copy from Java.Timer on what the behavior would be of the variables.

File: extensions/jdbc-store/src/main/java/com/cognitree/kronos/scheduler/store/jdbc/StdJDBCStoreService.java (hunk 1 above)
Comment (ankitnanglia) [reply extensions/jdbc-store/src/main/java/com/cognitree/kronos/scheduler/store/jdbc/StdJDBCStoreService.java]: the class is supposed to be implemented by the user. User might not remember to register the service. Have moved the logic to register the store to StoreService.

File: extensions/mongo-store/README.md (hunk 2 above)
Comment (ankitnanglia) [reply extensions/mongo-store/README.md]: quartz stores its data in a different database. Randomly picking a name for storing quartz configuration can be tricky and error-prone.
Although in StdJDBCStore we append all the quartz table `QRTZ` which is fixed as we do not have control to select the table name explicitly.

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoStoreService.java (hunk 10 above)
Comment (ankitnanglia) [reply extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoStoreService.java]: the constants are named like this across Kronos. We can decide to update them all or leave it as it is.

Is this causing ambiguity?

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java (hunk 12 above)
Comment (ankitnanglia) [reply extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java]: TaskStore and JobStore are two different interfaces altogether. Not sure what you mean by merging them with generics. 

Very little can be extracted out in a helper class which did not make sense to me

Comment (ankitnanglia) [reply extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoTaskStore.java]: An example will help

File: extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java (hunk 8 above)
Comment (ankitnanglia) [reply extensions/mongo-store/src/main/java/com/cognitree/kronos/scheduler/store/mongo/MongoJobStore.java]: we are doing aggregations here 
And it will have it's own format
